### PR TITLE
#257: create new id_v2 option

### PR DESCRIPTION
**Pull request by mmezhensky**
### Description (*)

This PR extends product schema and added new identifier "**id_v2**"

[Solution Architecture](https://github.com/magento/architecture/blob/master/design-documents/graph-ql/coverage/add-items-to-cart-single-mutation.md#option-implementation)

### Fixed Issues (if relevant)

1. [Single mutation for adding products to cart](https://github.com/magento/community-features/issues/257)

### Manual testing scenarios (*)

```text
query {
  products(filter: { price: { from: "0.00001" } }) {
    items {
      sku

      ... on CustomizableProductInterface {
        options {
          option_id
          title
          
          ... on CustomizableRadioOption {
            title
            option_id
            selected_option: value {
              option_type_id
              uid
            }
          }
          
          ... on CustomizableDropDownOption {
            title
            option_id
            selected_option: value {
              option_type_id
              uid
            }
          }
          
          ... on CustomizableMultipleOption {
            option_id
            title
            selected_option: value {
              option_type_id
              uid
            }
          }
          
          ... on CustomizableCheckboxOption {
            option_id
            title
            selected_option: value {
              option_type_id
              uid
            }
          }
          
          ... on CustomizableAreaOption {
            option_id
            title
            entered_option: value {
              uid
            }
          }
          
          ... on CustomizableFieldOption {
            option_id
            title
            entered_option: value {
              uid
            }
          }
          
          ... on CustomizableFileOption {
            title
            option_id
            entered_option: value {
              uid
            }
          }
            
          ... on CustomizableDateOption {
            title
            option_id
            entered_option: value {
              uid
            }
          }
        }
      }
      
      ... on ConfigurableProduct {
        variants {
          attributes {
            label
            value_index
            uid
          }
        }
      }
      
      ... on DownloadableProduct {
        downloadable_product_links {
          title
          uid
        }
      }
      
      ... on BundleProduct {
        items {
          title
          options {
            uid
          }
        }
      }
    }
  }
}
```

### Questions or comments
⚠️ For your information, that we have to use aliases (`selected_options`, `entered_options`), because both cases have a conflicting return type for `value`.

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds are green)

### CODE REVIEW

**Comment by eduard13** [comment]
@magento run Static Tests

**Comment by eduard13** [comment]
@magento run Static Tests

**Comment by eduard13** [comment]
@magento run all tests

File: app/code/Magento/ConfigurableProductGraphQl/Model/Resolver/Variant/Attributes/ConfigurableAttributeIdV2.php
Comment (rogyar) [inline app/code/Magento/ConfigurableProductGraphQl/Model/Resolver/Variant/Attributes/ConfigurableAttributeIdV2.php]: I believe `$attribute_id` is supposed to be a part of the `$value` array but not a standalone variable

**Review by rogyar — CHANGES_REQUESTED** [review]

**Comment by rogyar** [comment]
Looks good to me. Let's wait for the final decision regarding the bundle product options with quantity

**Review by rogyar — DISMISSED** [review]

**Comment by magento-engcom-team** [comment]
Hi @rogyar, thank you for the review.
[ENGCOM-7608](https://jira.corp.magento.com/browse/ENGCOM-7608) has been created to process this Pull Request

File: app/code/Magento/BundleGraphQl/Model/Resolver/Options/BundleEnteredOptionValueIdV2.php
Comment (prabhuram93) [inline app/code/Magento/BundleGraphQl/Model/Resolver/Options/BundleEnteredOptionValueIdV2.php]: Implementation looks neat. The name here BundleEnteredOptionValueIdV2 is little confusing because a bundle option is something that need not be entered. lets just keep it simple and name it BundleItemOptionIdV2, the same hierarchy it follows in the schema.

File: app/code/Magento/ConfigurableProductGraphQl/Model/Resolver/Variant/Attributes/ConfigurableAttributeIdV2.php
Comment (prabhuram93) [inline app/code/Magento/ConfigurableProductGraphQl/Model/Resolver/Variant/Attributes/ConfigurableAttributeIdV2.php]: Using eavAttribute adds a hard dependency. app/code/Magento/ConfigurableProductGraphQl/Model/Resolver/Variant/Attributes.php:47 here the $value['options'] has the attribute_id which we need. Just make sure to return it along with the existing data that is being returned ('code' and other stuff) so you can leverage 'attribute_id' it on the ConfigurableAttributeIdV2 resolver. This will remove the hard dependency.

File: app/code/Magento/ConfigurableProductGraphQl/composer.json
Comment (prabhuram93) [inline app/code/Magento/ConfigurableProductGraphQl/composer.json]: This can be avoided based on the above comment.

File: app/code/Magento/CatalogGraphQl/etc/schema.graphqls
Comment (prabhuram93) [inline app/code/Magento/CatalogGraphQl/etc/schema.graphqls]: @keharper Please review the schema descriptions.

**Review by prabhuram93 — CHANGES_REQUESTED** [review]

File: app/code/Magento/CatalogGraphQl/etc/schema.graphqls
Comment (keharper) [reply app/code/Magento/CatalogGraphQl/etc/schema.graphqls]: @mmezhensky Change the descriptions of `id_v2` to "A base64 string that encodes option details."

Comment (DrewML) [reply app/code/Magento/CatalogGraphQl/etc/schema.graphqls]: IDs should be opaque to a client - let's avoid putting anything about base64 encoding into the docs for ID fields - it's an implementation detail.

It can be an inline comment in the `.graphqls` file with the `#` character as a hint to developers working on the implementation, though.

File: app/code/Magento/BundleGraphQl/Model/Resolver/Options/BundleEnteredOptionValueIdV2.php
Comment (eduard13) [reply app/code/Magento/BundleGraphQl/Model/Resolver/Options/BundleEnteredOptionValueIdV2.php]: Agree on that, the new name seems much better. However, we may also have the possibility to enter the option's quantity, which makes the quantity an entered value (I've also raised this question in slack channel).
Thank you for the suggestion.

File: app/code/Magento/ConfigurableProductGraphQl/Model/Resolver/Variant/Attributes/ConfigurableAttributeIdV2.php
Comment (eduard13) [reply app/code/Magento/ConfigurableProductGraphQl/Model/Resolver/Variant/Attributes/ConfigurableAttributeIdV2.php]: That makes sense. Thank you for this one too, totally agree 👍.

File: app/code/Magento/CatalogGraphQl/etc/schema.graphqls
Comment (eduard13) [reply app/code/Magento/CatalogGraphQl/etc/schema.graphqls]: @keharper, @DrewML the description were updated. The Base64 encoding is mentioning as field comment.

**Comment by eduard13** [comment]
Hi @prabhuram93, thank you for reviewing this one, could you please have a look at the updated changes, so we can move forward with this one?
Thank you.

**Comment by eduard13** [comment]
@magento run WebAPI tests, Static Tests, Unit Tests

**Comment by eduard13** [comment]
@magento run WebAPI Tests

**Comment by eduard13** [comment]
@magento run all tests

**Review by prabhuram93 — DISMISSED** [review]

**Comment by prabhuram93** [comment]
> Hi @prabhuram93, thank you for reviewing this one, could you please have a look at the updated changes, so we can move forward with this one?
> Thank you.

Good with the changes. Approved it 👍

**Comment by dthampy** [comment]
Reviewed the tests and looks good. Testing complete.

**Comment by prabhuram93** [comment]
@eduard13 @rogyar I see  [Single mutation for adding products to cart](https://github.com/magento/community-features/issues/257) in the Fixed issues on this ticket. is it relevant?

**Comment by DrewML** [comment]
Note: We can update the architecture document, but occurrences of `id_v2` should be an `ID` type, not a `String`. This allows us to change the format of the ID later, and also signals to developers that they're getting an opaque value, rather than a string they can perform some operations on.

**Comment by prabhuram93** [comment]
> Note: We can update the architecture document, but occurrences of `id_v2` should be an `ID` type, not a `String`. This allows us to change the format of the ID later, and also signals to developers that they're getting an opaque value, rather than a string they can perform some operations on.
> 
> [We're finishing up formalizing these guidelines](https://github.com/magento/architecture/pull/395), and would be great to get this in

On the same lines, ID should be a mandatory field. Would be great if we can update this on the schema.

**Comment by eduard13** [comment]
@magento run WebAPI Tests, Static Tests, Integration Tests, Unit Tests

**Comment by eduard13** [comment]
@magento run all tests
Hi @prabhuram93, @DrewML thank you for the suggestions, the implementation has been adjusted by renaming it to `uid` and making it a mandatory field. Please let me know if I missed anything.
Thank you.

**Review by prabhuram93 — APPROVED** [review]

**Comment by prabhuram93** [comment]
> @magento run all tests
> Hi @prabhuram93, @DrewML thank you for the suggestions, the implementation has been adjusted by renaming it to `uid` and making it a mandatory field. Please let me know if I missed anything.
> Thank you.

Looks good! Thank you @eduard13

**Review by dthampy — APPROVED** [review]

**Pull request merged**
